x/dex/exchange/settlement.go

```diff
@@ -94,24 +111,40 @@ func SettleFromBook(
 	}
 	longOrder.GetEntry().Quantity = longOrder.GetEntry().Quantity.Sub(executedQuantity)
 	shortOrder.GetEntry().Quantity = shortOrder.GetEntry().Quantity.Sub(executedQuantity)
-	newLongAllocations := RebalanceAllocations(longOrder)
-	newShortAllocations := RebalanceAllocations(shortOrder)
 	newLongToSettle := []ToSettle{}
 	newShortToSettle := []ToSettle{}
 	nonZeroNewLongAllocations, nonZeroNewShortAllocations := []*types.Allocation{}, []*types.Allocation{}
+	longSettledQuantity, shortSettledQuantity := sdk.ZeroDec(), sdk.ZeroDec()
 	for _, allocation := range longOrder.GetEntry().Allocations {
 		if allocation.Quantity.IsZero() {
 			continue
 		}
-		newLongToSettle = append(newLongToSettle, ToSettle{
-			amount:  allocation.Quantity.Sub(newLongAllocations[allocation.OrderId]),
-			account: allocation.Account,
-			orderID: allocation.OrderId,
-		})
-		if newLongAllocations[allocation.OrderId].IsPositive() {
+		if longSettledQuantity.LT(executedQuantity) {
+			diff := executedQuantity.Sub(longSettledQuantity)
+			if allocation.Quantity.GT(diff) {
+				newLongToSettle = append(newLongToSettle, ToSettle{
+					amount:  diff,
+					orderID: allocation.OrderId,
+					account: allocation.Account,
+				})
+				nonZeroNewLongAllocations = append(nonZeroNewLongAllocations, &types.Allocation{
+					OrderId:  allocation.OrderId,
+					Quantity: allocation.Quantity.Sub(diff),
+					Account:  allocation.Account,
+				})
+				longSettledQuantity = executedQuantity
+			} else {
+				newLongToSettle = append(newLongToSettle, ToSettle{
+					amount:  allocation.Quantity,
+					orderID: allocation.OrderId,
+					account: allocation.Account,
+				})
+				longSettledQuantity = longSettledQuantity.Add(allocation.Quantity)
+			}
+		} else {
 			nonZeroNewLongAllocations = append(nonZeroNewLongAllocations, &types.Allocation{
```

Comment:
do we need to break here? it seems that we are still going over the entire long book allocations even if we have fulfilled the amount of `executedQuantity` here

Reply:
actually is it on purpose? why do we create a new long allocations instead of modifying the existing one here

Reply:
it's intentional. I don't think a go slice can be modified and we would need to create a new slice if we want to delete items from a slice